// sealights-cypress-plugin/tsOutputs/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TEST_STATUS = exports.SL_COLORING_HEADERS = exports.SL_COLORING_EVENTS = exports.SL_REPORT_TEST_END_TASK = exports.X_SL_SERVER = exports.SL_TASKS_REGISTERED = exports.SL_END_TEST_SESSION = exports.SL_START_TEST_SESSION = exports.SL_LOG_PREFIX = exports.SL_CYPRESS_ENVIRONMENT = void 0;
exports.SL_CYPRESS_ENVIRONMENT = {
    BUILD_SESSION_ID: 'SL_BUILD_SESSION_ID',
    TOKEN: 'SL_TOKEN',
    TEST_STAGE: 'SL_TEST_STAGE',
    LAB_ID: 'SL_LAB_ID',
};
exports.SL_LOG_PREFIX = 'Sealights Plugin:';
exports.SL_START_TEST_SESSION = 'sealightsStartSession';
exports.SL_END_TEST_SESSION = 'sealightsEndSession';
exports.SL_TASKS_REGISTERED = 'sealightsTasksRegistered';
exports.X_SL_SERVER = 'x-sl-server';
exports.SL_REPORT_TEST_END_TASK = 'reportSealightsTestEnd';
exports.SL_COLORING_EVENTS = {
    SET_CONTEXT: 'set:context',
    DELETE_CONTEXT: 'delete:context',
};
exports.SL_COLORING_HEADERS = {
    TEST_NAME: 'x-sl-test-name',
    TEST_SESSION_ID: 'x-sl-test-session-id',
};
var TEST_STATUS;
(function (TEST_STATUS) {
    TEST_STATUS["SKIPPED"] = "skipped";
    TEST_STATUS["PENDING"] = "pending";
})(TEST_STATUS || (exports.TEST_STATUS = TEST_STATUS = {}));
